[StructuralMechanics] Updating AnalysisStage Class

philbucher opened the pull request:
This PR updates the AnalysisStage Class of StructuralMechanics, as previously discussed
Minor Bugfixes in the BaseClass are also there


see #1904 and #1886 

@KratosMultiphysics/technical-committee note that without #1886 I cannot use the `AnalysisStage` without calling `RunSolutionLoop` (unless I call e.g. `Stage.solver.Predict`)
This would be needed if in a standard analysis I want to print some values after `Predict`

philbucher committed:
Renamed AnalysisStage

philbucher committed:
More naming changes

philbucher committed:
Now the Analysis derives from the Core one

philbucher committed:
Moved selection of Logger severity to Tests (where it belongs)

philbucher committed:
bugfixes in BaseAnalysis

philbucher committed:
Moved process construction to Function

philbucher committed:
Cleanup

philbucher committed:
Introducing OutputSolutionStep

philbucher committed:
Using the calls to the Strategy properly

philbucher committed:
Using Finalize Properly

philbucher committed:
Implemented AdvanceInTime

philbucher committed:
Cleanup

philbucher committed:
Cleanup II

philbucher commented:
@KratosMultiphysics/optimization the interface changed if you want to use an external Modelpart,
Please talk to me, explaining it in person is easier than here

loumalouomega commented:
I need to take a look if I need to modify something in my contact analysis

El vie., 20 abr. 2018 10:56, Philipp Bucher <notifications@github.com>
escribió:

> @KratosMultiphysics/optimization
> <https://github.com/orgs/KratosMultiphysics/teams/optimization> the
> interface changed if you want to use an external Modelpart,
> Please talk to me, explaining it in person is easier than here
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/1909#issuecomment-383032160>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATEMgAkOV_cU8TqzKXmoQDsRBs2XWAuyks5tqaK8gaJpZM4TdGow>
> .
>

philbucher commented:
@loumalouomega I suggets you to wait until the others gave their comments, then you don't have to do it twice

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -124,6 +124,13 @@ def __init__(self, main_model_part, custom_settings):
         else:
             self.main_model_part.ProcessInfo[KratosMultiphysics.IS_RESTARTED] = False
 
+        # This is needed when the solving functions are called, only NewtonRaphson properly implements
```
we should do the change in the linear_strategy too...

philbucher replied:
absolutely, this is why I opened #1911

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_analysis_stage.py:
```diff
@@ -16,95 +16,115 @@
 except ImportError:
     KratosMultiphysics.Logger.PrintInfo("EigenSolversApplication", "not imported")
 
+# Importing the base class
```
a Few questions:

shouldn't InitializeIO, ExecuteInitialize go in the base class? i guess they are the same for fluid and structure.

PrepareModelPartForSolver is a new function of the Solver API? i mean all solvers will have to provide that?

philbucher replied:
eventually these will go the the base class, yes!

Then it will be easier to do only solver-coupling, what you want

We could do this now, but then we will need another PR for updating the baseclass of the AnalysisStage

The `PrepareModelPartForSolver` was introduced by me for @KratosMultiphysics/optimization because they pass an externally read modelpart. Therefore I separated `ImportModelPart` into `ReadModelPart` and `PrepareModelPartForSolver` (which does the filling of the buffer and the creation of the ComputingModelPart)

RiccardoRossi replied:
Ok, this makes sense to me. I guess it should go in the API of the solver

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_analysis_stage.py:
```diff
@@ -140,42 +159,37 @@ def _ExecuteInitialize(self):
+    def _SetUpListOfProcesses(self):
+        from process_factory import KratosProcessFactory
+        factory = KratosProcessFactory(self.model)
+        self.list_of_processes = factory.ConstructListOfProcesses(self.project_parameters["constraints_process_list"])
```
do we need to distinguish the origin? i mean "constraints_process_list", loads_process_list, etc. ?

Ifthis is not so, then it would be better to have everythign in a single list, since anyway wethen push it back to a single list of processes

philbucher replied:
@RiccardoRossi they don't need to be in different ones (even though I find it good to destinguish what a process is being used for)

=> I don't know who and why did this
I talked to @jcotela and we think that if we do not destinguish the origin, then it could be put to the base class

Note that this BREAKS the current Interface, i.e. old cases will stop working!

If you are willing to pay this price we can do it

RiccardoRossi replied:
I don't think this is my call... in the sense that i should not be the one to decide this. Thing is that if the goal is unifying than we should go that way...or decide not to, preferably for good technical reasons

philbucher commented:
@RiccardoRossi and @KratosMultiphysics/technical-committee 
From the technical point of view I am quite convinced that we can unify the processes, since as you see, even now they go to the same variable (`list_of_processes`)

Furthermore we don't necessarily have to break the interface, we can have an intermediate solution, similar what we do in the structural solver with deprecated settings. It would still work but we would throw a warning abt how things have to be changed. (Sorry for changing my mind but this came to my mind only yesterday)

the structure in the ProjectParams could look like this:
~~~
"processes" : [
    "process_1" : { 
    ...
    },
    "process_2" : {
    ...
    }
]
~~~
With this structure one looses the overview a bit, in the sense that it is messy and one does not know which processes belong to BCs, loads, other things etc.
##
Therefore the following structure could work: (wrapping the current processes in a "processes" block):
~~~
"processes" : [
    "load_process" : { 
        "force_support" : {
        ...
        }
    },
    "initial_conditions" : {
        process_1 : {
        ...
        },
        process_2 : {
        ...
        }
    }
]
~~~
Constructing the processes is easy in both cases!

In the second one the user even has the option too do sth special with a certain set of processes in a custom Stage. Plus this can be used bcs I think in fluids the order of initialization for the Processes matters (Initial Conditions before Boundary conditions right @jcotela ?)

**Summing it up**:
If we have this then a lot of things can go to the base-class of the Stage!!! The ones in Structural and Fluid will be small (essentially importing the correct solver which is in the Apps) + in Fluid the corrected list of Process initialization (what I wrote above)
The remaining differences we can also sort out, there are I think no technical reasons other things are done differently

jcotela commented:
(posting here instead of in the fluid PR to unify the discussion)
@RiccardoRossi I remember discussing with @rubenzorrilla at some point that the processes in the fluiddynamics are separated because the order in which they are applied is important. (you don't want the initial conditions to overwrite boundary conditions, the outlet condition for FS needs to have the gravity pre-defined to apply the correct boundary force, etc.). I don't recall the specifics, but I think this is the technical reason why processes are separated in the fluid application.

philbucher commented:
@RiccardoRossi where would the importing of the Applications go?
To the Stage or to the Main?

Also @RiccardoRossi do you have time to join the Skype tmr at 11am?

RiccardoRossi commented:
I agree that ordering matters. Then i think we cannot unify this...

i will make my best to attend tomorrow

philbucher committed:
Merge branch 'master' into StructMech/analysis-stage-update

philbucher committed:
renamed to be consistent with Fluid

philbucher committed:
Final Updates

philbucher commented:
This is ready to be re-reviewed

I didn't make any changes to the baseclass, in fact this should only contain the changes that everyone was ok with

@RiccardoRossi @jcotela could you please have a look?
Ideally this will be merged before the release branch is made such that we can also update the interface

philbucher committed:
re-adding missing fct

jcotela commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_analysis.py:
```diff
@@ -16,95 +16,106 @@
+
+        if self.is_printing_rank:
+            KratosMultiphysics.Logger.PrintInfo("STEP: ", self.main_model_part.ProcessInfo[KratosMultiphysics.STEP])
+            KratosMultiphysics.Logger.PrintInfo("TIME: ", self.time)
```
You may want to add a label here.

jcotela commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_analysis.py:
```diff
@@ -16,95 +16,106 @@
-    def Finalize(self):
-        self._ExecuteFinalize()
+        if (self.output_post == True):
+            self.gid_output.ExecuteFinalizeSolutionStep()
```
You can add the output to the standard list of processes and you won't need to redefine this function here (this also ensures that all standard process methods in gid_output_process are called at the right time)

jcotela reviewed (dismissed):
Approving, with a couple of minor comments.

philbuchersimscale committed:
Addressed Jordis comments

philbuchersimscale committed:
Cleanup

philbucher commented:
@jcotela I addressed your comments, simplifed everything even further

Btw the fluidanalysis will fail currently if you don't specify an output_configuration in the json, but this should not be part of this PR

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_analysis.py:
```diff
@@ -113,89 +113,81 @@ def _CreateSolver(self, external_model_part=None):
-        """ Initialize GiD  I/O """
-        self.output_post  = self.ProjectParameters.Has("output_configuration")
-        if (self.output_post == True):
+    def _SetUpGiDOutput(self):
```
What if we wanted to export to hdf5 ? the name does not suggest it

philbucher replied:
I discussed this with @jcotela , this will be improved in a later PR

We have sth more general in mind that will be in the baseclass

jcotela replied:
Then we will have a proper factory instead of this function (that is the plan for the fluid, at least...).

RiccardoRossi reviewed (approved):
this looks ok to me. I am approving on my side (apart for a minor comment on the name of GiD output not being appropriate for HDF5, but this can be fixed in a later PR)

@jcotela accept you too if you are ok
